### ISSUE TITLE
try_cast to TIMESTAMP reports compilation error

### ISSUE THREAD

**Issue by xerial** [facebook/presto#2380]
In 0.94-SNAPSHOT (we are using this version: facebook/presto@39202d3ffeb50897c0d7d8a30c29c13a01773dd3), try_cast to TIMESTAMP fails, while cast succeeds. 
```
presto:support> select cast(a as timestamp) from (VALUES ('2015-02-19')) as t(a);
          _col0
-------------------------
 2015-02-19 00:00:00.000
(1 row)
Query 20150219_044448_01327_739kf, FINISHED, 1 node
Splits: 1 total, 1 done (100.00%)
0:00 [0 rows, 0B] [0 rows/s, 0B/s]
```

```
presto:support> select try_cast(a as timestamp) from (VALUES ('2015-02-19')) as t(a);
Query 20150219_044453_01330_739kf failed: Compiler failed and interpreter is disabled
```

Here is a part of the stack trace:
```
"failureInfo": {
"type": "com.facebook.presto.spi.PrestoException",
"message": "Compiler failed and interpreter is disabled",
"cause": {
"type": "com.google.common.util.concurrent.UncheckedExecutionException",
"message": "java.lang.invoke.WrongMethodTypeException: cannot convert MethodHandle(ConnectorSession,Slice)long to (Slice)Long",
"cause": {
"type": "java.lang.invoke.WrongMethodTypeException",
"message": "cannot convert MethodHandle(ConnectorSession,Slice)long to (Slice)Long",
"suppressed": [],
"stack": [
"java.lang.invoke.MethodHandle.asTypeUncached(MethodHandle.java:776)",
"java.lang.invoke.MethodHandle.asType(MethodHandle.java:770)",
"com.facebook.presto.operator.scalar.TryCastFunction.specialize(TryCastFunction.java:83)",
"com.facebook.presto.metadata.FunctionRegistry$1.load(FunctionRegistry.java:217)",
"com.facebook.presto.metadata.FunctionRegistry$1.load(FunctionRegistry.java:212)",
"com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3527)",
"com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2319)",
"com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2282)",
"com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2197)",
"com.google.common.cache.LocalCache.get(LocalCache.java:3937)",
"com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3941)",
"com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4824)",
"com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4830)",
"com.facebook.presto.metadata.FunctionRegistry.getExactFunction(FunctionRegistry.java:477)",
"com.facebook.presto.sql.gen.FunctionCallCodeGenerator.generateExpression(FunctionCallCodeGenerator.java:36)",
"com.facebook.presto.sql.gen.ByteCodeExpressionVisitor.visitCall(ByteCodeExpressionVisitor.java:110)",
"com.facebook.presto.sql.gen.ByteCodeExpressionVisitor.visitCall(ByteCodeExpressionVisitor.java:40)",
"com.facebook.presto.sql.relational.CallExpression.accept(CallExpression.java:87)",
"com.facebook.presto.sql.gen.PageProcessorCompiler.generateProjectMethod(PageProcessorCompiler.java:267)",
"com.facebook.presto.sql.gen.PageProcessorCompiler.generateMethods(PageProcessorCompiler.java:77)",
"com.facebook.presto.sql.gen.ExpressionCompiler.compileProcessor(ExpressionCompiler.java:119)",
"com.facebook.presto.sql.gen.ExpressionCompiler.compileAndInstantiate(ExpressionCompiler.java:95)",
"com.facebook.presto.sql.gen.ExpressionCompiler.access$300(ExpressionCompiler.java:44)",
"com.facebook.presto.sql.gen.ExpressionCompiler$1.load(ExpressionCompiler.java:55)",
"com.facebook.presto.sql.gen.ExpressionCompiler$1.load(ExpressionCompiler.java:50)",
"com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3527)",
"com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2319)",
"com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2282)",
"com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2197)",
"com.google.common.cache.LocalCache.get(LocalCache.java:3937)",
"com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3941)",
"com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4824)",
"com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4830)",
"com.facebook.presto.sql.gen.ExpressionCompiler.compilePageProcessor(ExpressionCompiler.java:89)",
"com.facebook.presto.sql.planner.LocalExecutionPlanner$Visitor.visitScanFilterAndProject(LocalExecutionPlanner.java:811)",
"com.facebook.presto.sql.planner.LocalExecutionPlanner$Visitor.visitProject(LocalExecutionPlanner.java:724)",
"com.facebook.presto.sql.planner.LocalExecutionPlanner$Visitor.visitProject(LocalExecutionPlanner.java:378)",
"com.facebook.presto.sql.planner.plan.ProjectNode.accept(ProjectNode.java:81)",
"com.facebook.presto.sql.planner.LocalExecutionPlanner$Visitor.visitOutput(LocalExecutionPlanner.java:401)",
"com.facebook.presto.sql.planner.LocalExecutionPlanner$Visitor.visitOutput(LocalExecutionPlanner.java:378)",
"com.facebook.presto.sql.planner.plan.OutputNode.accept(OutputNode.java:79)",
"com.facebook.presto.sql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:222)",
```

**Comment by electrum** [comment]
Nice find! The method handle code in `TryCastFunction` needs the same behavior as `ByteCodeUtils.generateInvocation()` to handle a `ConnectorSession` argument. Though I'm not sure that is possible to do currently with `ParametricScalar`.

**Comment by electrum** [comment]
Fixed in #2420
